`Notifications`: Add basic weekly summaries

Malyuk-A opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [ ] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [ ] I implemented the changes with a good performance and prevented too many database calls.
- [ ] I documented the Java code using JavaDoc style.
#### Client
- [ ] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

### Description
<!-- Describe your changes in detail -->

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Log in to Artemis
2. Navigate to Course Administration
3. ...

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Please use the schema "Branches % | Lines %" -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- - ExerciseService.java: 85% | 77% -->
<!-- - programming-exercise.component.ts: 13% | 95% -->

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

willergeorg reviewed (dismissed):
Tested locally and worked fine. After discussing it with Alex, code changes look reasonable to me as well.

filip-gregurevic reviewed (dismissed):
Looks good to me 🚀

willergeorg reviewed (approved):
Changes since last review look good. Re-approve.

b-fein commented on src/main/java/de/tum/in/www1/artemis/service/EmailSummaryService.java:
```diff
@@ -0,0 +1,156 @@
+        if (exercise == null || exercise.getReleaseDate() == null) {
+            return false;
+        }
+        return exercise.isReleased() && exercise.getReleaseDate().isAfter(ZonedDateTime.now().minus(scheduleInterval)) && !exercise.isEnded();
```
This will break when merging, as the `exercise.inEnded()` method no longer exists.

I think just a simple check if `ZonedDateTime.now().isBefore(exercise.getDueDate())` should be enough. Alternatively, if you care about the individual due dates, you can use `ExerciseDateService::isBeforeLatestDueDate`.

Malyuk-A replied:
Thank you very much for pointing this out. I replaced the method call with your first recommendation. 👍

Malyuk-A commented:
As @b-fein mentioned, the latest commit only changes one (soon to be outdated/unavailable) method call in 1 line.
This PR is still mergeable (it had/has 3 approvals)

willergeorg reviewed (approved):
Code changes since last review seem reasonable to me.

Malyuk-A commented:
The one codacy issue is due to a very generic exception that might be caught. I followed the convention of all other scheduling services here so I think it is ok.

willergeorg reviewed (dismissed):
Re-tested locally and still works fine. Code changes since last review seem to be reasonable as well.

filip-gregurevic reviewed (dismissed):
Looks good to me 👍🏼

DanielKainz reviewed (approved):
Tested it. Everything worked. Great work :-)

krusche reviewed (changes requested):
The code in this PR is a performance nightmare for large Artemis instances with many users. It will definitely kill the database and the Artemis instance which tries to send the emails.

I do not understand how it can get 3 approvals and how you can blindly tick off the task

- [ ]  I implemented the changes with a good performance and prevented too many database calls.

which is definitely NOT FULFILLED at all.

Here is some detailed feedback:

We should disable the Weekly Summary by default, so that users only get it when they enable it. It is not feasible to send emails to all users (in the TUM instance we have 18,600), which is currently implemented very inefficiently. We should not load all users and then make a DB call for each user to check whether he or she wants the summary, because the production server or the DB would get too high load and die in doubt. So, the first call would have to load only the users who have activated the weekly summary in their settings. The check has to be done in HQL.

Also the creation of the email is implemented very inefficient. We would have to send 18,600 complex DB calls and load a lot of data multiple times, although this would be unnecessary. 
Here, we would have to load the relevant exercises from the database in one call (all exercises that have been released in the last 7 days and are still open, which are not so many in the normal case, probably between 0 and max. 50 in large Artemis instances). Then, we would have to filter for each user the exercise from the database call, for which the user is student in the course. This can be done well in Java and would not require much performance / memory.

Malyuk-A commented:
I implemented the requested changes (concerning the retrieval of the data (e.g. DB) -> there were no other change requests (so far), therefore the testing should be quite easy (and code review is focused on the new changes, which are quite small but very important in regards to performance)

The biggest change is that I load/filter all users that should be notified (i.e. have their weekly summary NotificationSetting enabled via a SQL query. (For comprehension: Each NotificationSetting has a reference to the respective User -> Due to Hibernate's lazy evaluation I have to fetch the needed user properties (e.g. group information for later) but it is not possible to only select the users in one query, therefore the other method that extracts the users is needed. (I discussed this query with Stephan and we think this is ok).
![image](https://user-images.githubusercontent.com/65814168/148373677-cd8e6192-9c06-4a41-a79a-932906fd5044.png)

The needed exercises for the emails are now also loaded via one DB call (instead for each user individually). 
![image](https://user-images.githubusercontent.com/65814168/148375492-b167afdd-6a6a-45e5-a7b9-19d447f9dd7b.png)

Later on, I reuse this "pool" of exercises for each user and simply slice/filter the needed exercises by course.
![image](https://user-images.githubusercontent.com/65814168/148375214-126f8375-19c7-4554-bdeb-c46c197e3c11.png)

These changes should improve the performance and resolve the change requests.

+ I changed the weekly summary setting to be deactivated by default

willergeorg reviewed (approved):
After discussing some points with @Malyuk-A regarding the code changes they look good to me. As we now fetch all possibly relevant exercises for the emails and later on filter them for each user I think the implementation is more efficient now.

filip-gregurevic reviewed (approved):
This version of the implementation should definitely have a better performance than the prior one. Looks good to me, nice job 👍🏼

willergeorg reviewed (approved):
Code changes since last review look minor to me. Re-approve.

filip-gregurevic reviewed (dismissed):
Re-approve. Only merge-conflicts were resolved after my last review

cremertim reviewed (approved):
tested locally, everything worked as expected :) great work @Malyuk-A

filip-gregurevic reviewed (approved):
Re-approve, Code changes still look good to me, nice job 👍🏼

bassner reviewed (approved):
I've tested this according to the instructions above and it works well 👍

cremertim reviewed (approved):
Reapprove, after testing it locally again. Well done alex😊

akesfeden commented on src/main/java/de/tum/in/www1/artemis/service/EmailSummaryService.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    @Async
+    public void prepareEmailSummaries() {
+        checkSecurityUtils();
```
Is there a reason for calling `checkSecurityUtils()` method both here and at [`scheduleEmailSummaries()`](https://github.com/ls1intum/Artemis/pull/4512/files#diff-19fc1f340b973371be268212ba65f841d380a5d91015a1585fbaef58887b2cc4R83)? Are there different contexts involved?

Malyuk-A replied:
Good question,
I use this here because it is a method that is called asynchronously and from my experience, it will not work otherwise.
In the runnable function in scheduleEmailSummaries I simply followed other Runnables in our code stack : e.g. 
![image](https://user-images.githubusercontent.com/65814168/157914851-2fc82d40-92e8-4e24-b4f2-59fa0f0dd4ea.png)
![image](https://user-images.githubusercontent.com/65814168/157914855-801dac1b-2b08-4784-bddd-11893ae9dc48.png)
Because this runnable is the function that is called by the scheduler we need to get the authentication again (at least from my experience so far working with asyn and schedulers)

akesfeden replied:
Thank you for the detailed explanation, in that case everything looks good as far as I can see.

akesfeden reviewed (commented):
Tested by checking out the branch locally, works as described, opened Html email content printed to the console in a browser, looks nice (except for the Artemis logo not being found since its base url is hardcoded but this is not a change introduced in this PR). Code changes also look good, I have one question in a different comment that may be a minor redundancy but it can also be something I am missing. I will approve after we resolve it.

akesfeden reviewed (approved):
My question from my previous comment is answered, approving.

willergeorg reviewed (approved):
Tested again locally. Works well. 👍
